packages/app/pages/queue.tsx

```diff
@@ -0,0 +1,196 @@
+  const [queueId, setQueueId] = useState<number>(null);
+  const [questionDraftId, setQuestionDraftId] = useState<number>(null);
+  const [currentQuestion, setCurrentQuestion] = useState<Question>(null);
+
+  useEffect(() => {
+    if (profile) {
+      const selectedCourse: UserCourse = profile.courses[0];
+      setCourse(selectedCourse);
+      setQueueId(selectedCourse.course.id);
+    }
+  }, [profile]);
+
+  useEffect(() => {
+    if (queueId) {
+      getQuestions();
+    }
+  }, [queueId]);
+
+  /**
+   * Gets the questions for this course
+   */
+  const getQuestions = () => {
+    if (queueId) {
+      API.questions.index(queueId).then((q) => {
+        if (q) {
+          setQuestions(q);
+        }
+      });
+    }
+  };
+
+  const onOpenClick = useCallback((question: Question): void => {
+    setCurrentQuestion(question);
+    setOpenPopup(true);
+  }, []);
+
+  const onCloseClick = useCallback((): void => {
+    setOpenPopup(false);
+  }, []);
+
+  /**
+   * Student functions to support queue operations.
+   */
+
+  /**
+   * Creates a new Question draft for a student who has joined the queue.
+   */
+  const joinQueue = () => {
+    setIsJoining(true);
+
+    // API call to join queue, question marked as draft
+    API.questions.create(queueId, null).then((q) => {
+      if (q) {
+        // fetch updated question list
+        // store questionDraftId in state (should this be stored in state or fetched everytime using status.Drafting?)
+      }
+    });
+  };
+
+  const leaveQueue = () => {
+    setIsJoining(false);
+  };
+
+  /**
+   * Finishes creating a given question by updating the draft.
+   */
+  const finishQuestion = (text: string, questionType: QuestionType) => {
+    API.questions
+      .update(queueId, questionDraftId, {
+        text: text,
+        questionType: questionType,
+        status: OpenQuestionStatus.Queued,
+      })
+      .then((q) => {
+        if (q) {
+          // fetch updated question list
+          setIsJoining(false);
+        }
+      });
+  };
+
+  /**
+   * TA functions to support queue operations
+   */
+
+  /**
+   * Updates a given question to the given status.
+   * @param question the question being modified
+   * @param status the updated status
+   */
+  const updateQuestionTA = (question: Question, status: QuestionStatus) => {
+    API.questions
+      .update(queueId, question.id, {
+        status: status,
+      })
+      .then((q) => {
+        if (q) {
+          // fetch updated question list
+          getQuestions();
+          // fetch updated helping list
+          setCurrentQuestion(q);
+          // update helping state if none left
+        }
+      });
+  };
+
+  /**
+   * Sends a push notification to the student with the given Question
+   * @param question the question to be notified
+   */
+  const alertStudent = (question: Question) => {
+    // Send API request to trigger notification
+  };
+
+  return useMemo(() => {
```

Comment:
Good call on the `useMemo` here

Reply:
Honestly I don't know its mostly habit from work where we memoize everything to reduce re-rendering.

Reply:
Removed the useMemo calls for now, we agreed not to use them unless we have a noticeable performance drop